partition_snapshot_reader.hh

```diff
@@ -242,12 +244,12 @@ class partition_snapshot_flat_reader : public flat_mutation_reader::impl, public
             return in_alloc_section([&] () -> mutation_fragment_opt {
                 maybe_refresh_state(ck_range, last_row, last_rts);
 
-                position_in_partition::less_compare rt_less(_schema);
+                position_in_partition::less_compare rt_less(_query_schema);
                 while (has_more_range_tombstones()
                         && !rt_less(pos, peek_range_tombstone().position())
                         && (_rt_stream.empty() || !rt_less(_rt_stream.peek_next().position(), peek_range_tombstone().position()))) {
                     range_tombstone rt = pop_range_tombstone();
-                    if (rt.trim(_schema,
+                    if (rt.trim(_query_schema,
```

Comment:
Here `_query_schema` is used together with `ck_range`. `range_tombstone::trim` uses the schema to performs comparisons. Hence, we're using a comparator created from `_query_schema` to compare the bounds of `ck_range` with the range tombstone bounds.

On the other hand, in `do_refresh_state` you use `_snapshot_schema` together with `ck_range` too. A comparator created from `_snapshot_schema` is used to compare the bounds of `ck_range` with some other stuff.

During reversed queries, `_query_schema` and `_snapshot_schema` may use reversed orderings, so their comparators are reversed.

So it seems to me that at least one of these places are using the wrong comparator or the wrong range (something must be reversed). How can this work? Does this work?

Reply:
I wrote this comment looking at the "separate query_schema from _snapshot_schema" commit, but the following commit seems to answer my questions - you split `ck_range` into two as well.

Reply:
Well no, my questions are still now answered.

You only split `ck_range` in `next_range_tombstone` but not in `next_row`.
What is the range passed to `next_row`? Is it a "query range" or is it a "snapshot range"?
You're using `_query_schema` inside `next_row`, which suggests that you need to pass in a "query range".
However, at the call site in `read_next()` you pass `*_current_ck_range` whic his equal to `ck_range_snapshot`. So you pass in a "snapshot range".
That seems contradictory.

Reply:
next_row actually needs "snapshot range" and thus the code is correct as it is. I will change the naming.

Reply:
Then why does it use `_query_schema`? How come `_query_schema` gives the "right" comparisons for `ck_range_snapshot`?

Reply:
next_row uses query schema, but for different purpose

Reply:
next_row uses ck_range_snapshot to optionally refresh state (i.e. to fetch available row-ranges and range tombstone ranges).
However, for the rest of operation, it uses a comparator derived from query schema. Transition between the two lies in implementation of `rows_position` and the methods taking out from the range tombstone list (`pop_range_tombstone`, `peek_range_tombstone`).